### ISSUE TITLE
Updates on layouts UX

### ISSUE THREAD

**Issue by syl20bnr** [syl20bnr/spacemacs#4590]
After using it for the last months and given some feeback of the community here are the updates I want to make to the layout UX, some of them were in the original design of @CestDiego:

_The main goal of the UX is to be able to make layouts transparents for people not using them, even if they are always available and activated under the hood, so we want a seamless experience for both kind of users (layouts users and no-layout users)_

- `SPC b b`, `SPC TAB`, `SPC b n` and `SPC b p` are restricted to the current layout
- New key bindings `SPC b a` and `SPC b r` to add/remove buffers to current layout
- `SPC b a` list all the opened buffers, `SPC b r` list only the buffers of the current layout
- New key binding `SPC f a` synonym to `SPC f f` to be symmetric with `SPC b a`.
- New key binding `SPC B b` `SPC B n` and `SPC B p` corresponding to the lower case bindings but they don't add buffers to the current layout, and _ideally they also switch to the first layout where the buffer belongs_. The latter part is important to be able to get a more WM like UX, this is also a powerful way to combo buffer switch and layout switch in one command.
- The helm/ivy sources should display the layouts for each buffer.

Users not using layouts will always be in the default layouts which has a global scope, so all restricted key bindings act like they're expected to.

**Comment by Andre0991** [comment]
> ideally they also switch to the first layout where the buffer belongs
I second this. I myself have been trying to code this, by the way.

**Comment by agzam** [comment]
What about `SPC Tab`? would that be scoped to use only buffers in the current layout?

**Comment by syl20bnr** [comment]
Yes, `SPC tab` should be restricted as well, adding it.

**Comment by StreakyCobra** [comment]
:+1: Go for it

**Comment by jmiven** [comment]
:+1: I only recently began to use layouts but these improvements sound great. I've had issues remembering to use `SPC l b`.

**Comment by nashamri** [comment]
> `SPC b b`, `SPC TAB`, `SPC b n` and `SPC b p` are restricted to the current layout

Oh yes please!! :smile:

**Comment by CestDiego** [comment]
I've started to work on this in https://github.com/syl20bnr/spacemacs/pull/4771. Feel free to continue discussion :D By the way there is a weird bug with "SPC TAB" and I think I'll have to use my own version of `alternate buffer`

**Comment by syl20bnr** [comment]
Thank you @CestDiego :100:

**Comment by CestDiego** [comment]
@syl20bnr this is not yet complete. I still need to implement the two last points

**Comment by Andre0991** [comment]
@CestDiego Great work, I like the restricted version of `SPC b b`.

I have one question.

How do I know that a buffer will get associated with a layout?

For example, I'm using this predefined layout:

```
  (spacemacs|define-custom-layout "@Common-Lisp"
    :binding "c"
    :body
    (andre/layout-4-half-screen-left-and-three-right)
    (slime)
    (find-file "/Users/andre0991/Dropbox/ciencia_da_computacao/programming_languages/common_lisp/common_lisp.org")
    (find-file "/Users/andre0991/Dropbox/ciencia_da_computacao/programming_languages/common_lisp/tests.lisp")
    (eww "http://www.gigamonkeys.com/book/")

    (select-window-2)
    (switch-to-buffer "*slime-repl sbcl*")
    (select-window-3)
    (switch-to-buffer "common_lisp.org")
    (select-window-4)
    (switch-to-buffer "tests.lisp")
    (select-window-1)
    (switch-to-buffer "*eww*"))

```

The things opened with `find-files` are associated with this layout. However, `*eww*` and the slime repl don't (I think I'll use `persp-add-buffer` as a work around for now). So my question is: what's the condition that guarantees that a buffer will be part of a layout?

**Comment by CestDiego** [comment]
right now there is no guarantee, it's the same issue with ERC buffers not being added unless one adds a hook to it to add the buffer to the layout one is currently at.

**Comment by Andre0991** [comment]
Ah, ok. I'll use `persp-add-buffer` then.

**Comment by robbyoconnor** [comment]
What happened to `SPC bd`?

**Comment by cpaulik** [comment]
Is this new behaviour documented somewhere? I could not find it in the `spacemacs-layouts` README.

**Comment by sindikat** [comment]
### Question about `alternate-buffer-in-persp`
Behavior of <kbd>SPC TAB</kbd> (`alternate-buffer-in-persp`) is counter-intuitive, although kinda logical, and I don't like it. To reproduce:

1. Open some buffer, e.g. `file.txt`
2. Split window vertically: <kbd>SPC w /</kbd>
    - Now you have 2 windows, 1 and 2
3. Move cursor in window 1 but do not switch to window 2: <kbd>j j j j j</kbd>
    - You can see that in window 2 the deactivated cursor haven't moved, which is correct
4. Press <kbd>SPC TAB</kbd> twice
5. You are returned to `file.txt`, yet the cursor is not as on window 1 before, but is now as on window 2.

This actually makes sense in Emacs. When you have same buffer opened in 2 windows, it's logical to store different cursor positions for each window. For example, a person works on the same file, but on 2 different parts, so they open the buffer in 2 windows.

Yet when you have only 1 window for a buffer, you only need to store one cursor.

Yet in context of <kbd>SPC TAB</kbd> is counter-intuitive, because once I return to my buffer, I lose the cursor.

Is there any elegant solution to this problem?

#### System Info
- OS: gnu/linux
- Emacs: 24.5.2
- Spacemacs: 0.105.14
- Spacemacs branch: develop (rev. d52eb41)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(auto-completion better-defaults git github nixos org spell-checking syntax-checking themes-megapack elm emacs-lisp
                 (haskell :variables haskell-enable-ghc-mod-support t haskell-enable-hindent-style "johan-tibell")
                 idris latex markdown python racket rust)
```

### Related issue:
[`SPC TAB` alternate-buffer-in-persp is broken #5129](https://github.com/syl20bnr/spacemacs/issues/5129)

**Comment by nixmaniack** [comment]
@sindikat To modify different portions of the same file in split screen, I would recommend to take a look at `clone-indirect-buffer` and `clone-indirect-buffer-other-window`. It remembers cursor positions as you're expecting.

**Comment by sindikat** [comment]
@syl20bnr what do you think of <kbd>SPC b b</kbd>, <kbd>SPC TAB</kbd>, <kbd>SPC b n</kbd>, <kbd>SPC b p</kbd> incorporating indirect buffer concept, so that switching between buffers in the same window never loses the cursor? How hard would that be to implement?

**Comment by d12frosted** [comment]
Fixed with released of Spacemacs v0.200. Let us know if you still have any problems with this issue.
